js/objects/ExecutionStack.js

```
@@ -0,0 +1,93 @@
+/**
+ * ExecutionStack
+ * ---------------------------------
+ * I am an object that manages a stack of
+ * ActivationContext objects.
+ * I am designed to be used by System as the
+ * global execution stack.
+ *
+ */
+
+class ExecutionStack {
+    constructor(){
+        this._stack = [];
+        this._globals = {};
+
+        // Bound methods
+        this.pop = this.pop.bind(this);
+        this.push = this.push.bind(this);
+        this.setGlobal = this.setGlobal.bind(this);
+        this.getGlobal = this.getGlobal.bind(this);
+    }
+
+    pop(){
+        if(!this._stack.length){
+            return null;
+        }
+        return this._stack.pop();
+    }
+
+    push(anActivation){
+        this._stack.push(anActivation);
+    }
+
+    setGlobal(varName, value){
+        this._globals[varName] = value;
+    }
+
+    getGlobal(varName){
+        return this._globals[varName];
+    }
+
+    get current(){
+        if(!this._stack.length){
+            return null;
+        }
+        return this._stack[this._stack.length - 1];
+    }
+
+    get previous(){
+        if(!this._stack.length >= 2){
+            return null;
+        }
+        return this._stack[this._stack.length - 2];
+    }
+};
+
+class ActivationContext {
+    constructor(messageName, part, incomingMessage, handlerFunction){
+        this.part = part;
+        this.messageName = messageName;
+        this.message = incomingMessage;
+        this.handlerFunction = handlerFunction;
+        this._locals = {};
+
+        // Bound methods
+        this.get = this.get.bind(this);
+        this.getLocal = this.getLocal.bind(this);
+        this.setLocal = this.setLocal.bind(this);
+    }
+
+    get(varName){
+        let localValue = this.getLocal(varName);
+        if(localValue !== undefined){
+            return localValue;
+        };
+        // otherwise try to return a global
+        // variable
+        return window.System.executionStack.getGlobal(varName);
```

It looks like this will return `undefined` if it isn't set. Is that what we want? No error?

---

Yeah. The error is handled when the variable is looked up by the interpreter, [here](https://github.com/UnitedLexCorp/SimpleTalk/blob/master/js/ohm/interpreter-semantics.js#L861).

---

Hmm sorry I didn't fire up this branch and was only looking at the code. It looked like it just called the `ExecutionStack.getGlobal()` method in this file:

```
    getGlobal(varName){
        return this._globals[varName];
    }
```

which is why I thought it would just return undefined.

---

You are right though, at that point in the code it would return undefined. Perhaps we should throw these errors from here rather than in the interpreter (we would do the same for local var lookups in an ActiveContext). You think that's a better way to go?
  
Also I had provided a link at the wrong branch in my previous response. [This](https://github.com/UnitedLexCorp/SimpleTalk/blob/eric-stack/js/ohm/interpreter-semantics.js#L855) is the correct spot.

---

Ah I see that's why I was confused. Honestly I don't think it matters really as long as you're consistent. Handling it somewhere else is fine too. You don't need to move things around because I'm unfamiliar with the code.

---

Yeah I think at this point we should send/throw a `ReferenceNotKnown` message which can then be handled in our error message function, then marked up in the script editor (or later debugger)

---

I will add some official error types for variable lookup and push them to this branch